### PR TITLE
feat: OkHttp implementation for making HTTP calls and WebSocket connections

### PR DESCRIPTION
**Pull request by ttypic**
Resolves https://github.com/ably/ably-java/issues/120

## Release Notes

- **New Features**
	- Introduced support for using the Ably SDK under a proxy, with detailed instructions in the README.
	- Added a new `network-client-okhttp` module to enhance HTTP and WebSocket connections.
	- Implemented classes for HTTP calls and WebSocket clients using the OkHttp library.
	- Enhanced WebSocket functionality with ping listener support in the new classes.
	- Added a retry mechanism for tests to improve reliability.

- **Documentation**
	- Enhanced README with proxy configuration details and examples.

- **Chores**
	- Updated project settings to include the new `network-client-okhttp` module.
	- Added integration tests for OkHttp configurations in the CI workflow.

### CODE REVIEW

File: network-client-okhttp/src/main/java/io/ably/lib/network/OkHttpWebSocketEngineFactory.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package io.ably.lib.network;
+
+public class OkHttpWebSocketEngineFactory implements WebSocketEngineFactory {
+    @Override
```
Comment (sacOO7) [inline network-client-okhttp/src/main/java/io/ably/lib/network/OkHttpWebSocketEngineFactory.java:4]: Can we rename `WebsocketEngineFactory` to `WebsocketClientFactory`. This should also be reflected across all interface impl. `WebsocketEngineFactory` term feels confusing.

File: network-client-okhttp/src/main/java/io/ably/lib/network/OkHttpWebSocketEngine.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package io.ably.lib.network;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+
+public class OkHttpWebSocketEngine implements WebSocketEngine {
```
Comment (sacOO7) [inline network-client-okhttp/src/main/java/io/ably/lib/network/OkHttpWebSocketEngine.java:6]: Rename to 
`OkHttpWebSocketClient implements AblyWebsocketClient`

**Review by sacOO7 — CHANGES_REQUESTED** [review]
Can we add integration tests for OkHttp proxy? 
Also, it will be great if we can have OkHttp integration tests running ( for both rest and realtime ) by enabling some type of flag on CI ?

File: lib/src/main/java/io/ably/lib/transport/WebSocketTransport.java [hunk 3]
```diff
@@ -78,24 +107,7 @@
...
-                if (isTls) {
-                    SSLContext sslContext = SSLContext.getInstance("TLS");
-                    sslContext.init(null, null, null);
-                    SafeSSLSocketFactory factory = new SafeSSLSocketFactory(sslContext.getSocketFactory());
-                    configBuilder.sslSocketFactory(factory);
-                }
-
-                WebSocketEngine engine = engineFactory.create(configBuilder.build());
-                webSocketClient = engine.create(wsUri, new WebSocketHandler(this::receive));
+                webSocketClient = this.webSocketEngine.create(wsUri, new WebSocketHandler(this::receive));
```
[the remark this answers is not in the document]
Comment (ttypic) [reply lib/src/main/java/io/ably/lib/transport/WebSocketTransport.java:110]: this code is inside constructor, so no need to synchronize here

**Review by sacOO7 — APPROVED** [review]
LGTM, Btw, you can take a look at RabbitAI comments?
Some of those look sensible to implement

**Pull request merged**